---
title: "QCB455/COS551 Final Project Homology Heatmaps"
author: "Supraj Gunda"
output: pdf_document
urlcolor: blue
---

Homology heatmap was used in proposal but did not end up being used in final paper

imports
```{r message=FALSE}
library(circlize)
library(Biostrings)
library(ggplot2)
library(dplyr)
```

```{r}
# using the NCBI's BLAST tools, downloaded this from NCBI website and have it saved locally in my file structure
Sys.setenv(PATH = paste(Sys.getenv("PATH"), "~/qcb455/ncbi-blast/bin", sep = ":"))

# using these just for proof of concept, didn't end up using this visualization procedure because it took 10 hours to run BLAST on these two genomes (did it locally, so this might be why it took so long)
catGenome <- "./fastaFiles/GCF_018350175.1_F.catus_Fca126_mat1.0_genomic.fna"
mouseGenome <- "./fastaFiles/GCF_000001635.27_GRCm39_genomic.fna"

# running blast and saving the outputs
system(paste("makeblastdb -in", mouseGenome, "-dbtype nucl -out mouse_db"))
system(paste("blastn -query", catGenome, "-db mouse_db -out results.txt -outfmt 6"))
```

```{r}
blastResults <- read.table("results.txt", header = FALSE, sep = "\t", col.names = c("chr1", "chr2", "start1", "end1", "start2", "end2", "evalue", "bitscore", "alignment_length", "query_cov", "mismatch", "gap_open"))

# getting a table of all homologous regions and their locations in the two genomes
homologousRegions <- blast_results[, c("chr1", "start1", "end1", "chr2", "start2", "end2")]
```


```{r}
# find corresponding regions
filteredHomologousRegions <- homologousRegions %>%
  filter(grepl("^NC_", chr1) & grepl("^NC_", chr2))

# find how many homologous regions there are for each section of the genomes
homologyDensity <- filteredHomologousRegions %>%
  group_by(chr1, chr2) %>%
  summarise(homologousCount = n()) %>%
  ungroup()

# preprocessing using factor just to make plot look better
homologyDensity$chr1 <- factor(homologyDensity$chr1, levels = unique(filteredHomologousRegions$chr1))
homologyDensity$chr2 <- factor(homologyDensity$chr2, levels = unique(filteredHomologousRegions$chr2))

# using ggplot2 to make heatmap
heatmapPlot <- ggplot(homologyDensity, aes(x = chr1, y = chr2, fill = homologousCount)) +
  geom_tile(color = "white") +
  scale_fill_gradient(low = "lightblue", high = "darkblue", name = "Homology Density") +
  labs(title = "Homology Density Heatmap (Cat and Mouse Genomes)", x = "Cat Chromosomes", y = "Mouse Chromosomes") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(hjust = 0.5))

# show the heatmap
print(heatmapPlot)
```

```{r}
# save it as a pdf
ggsave("homology_heatmap.pdf", plot = heatmapPlot, width = 10, height = 8)
```
